core/src/main/java/org/infinispan/reactive/publisher/impl/PublisherHandler.java

```diff
@@ -495,187 +514,106 @@ public void run() {
-                  keyForSegmentCompletion = originalValue;
-
-                  ByRef.Integer size = new ByRef.Integer(0);
-                  return Flowable.fromPublisher(functionToApply.apply(Flowable.just(originalValue)))
-                        .doOnNext(ignore -> size.inc())
-                        .doOnComplete(() -> {
-                           int total = size.get();
-                           if (total > 0) {
-                              publisherOffset += total;
-                              // Means our values were consumed downstream immediately and thus our key is complete
-                              if (publisherOffset == consumerOffset) {
-                                 keyCompleted(originalValue);
-                                 previousValueFinishedKey = true;
-                              } else {
-                                 keyCompletionPosition.put(publisherOffset, originalValue);
-                              }
-                           } else {
-                              // If there are no values for the key it is complete but also doesn't need to
-                              // be tracked, so complete any segment tied to it if possible
-                              Integer segment = keySegmentCompletions.remove(originalValue);
-                              if (segment != null) {
-                                 if (log.isTraceEnabled()) {
-                                    log.tracef("Completing segment %s due to empty resulting value of %s for %s",
-                                          segment, originalValue, requestId);
-                                 }
-                                 actualCompleteSegment(segment);
-                              }
-                              // Also null out our key for segment completion if needed since this key will never be
-                              // found published
-                              if (keyForSegmentCompletion == originalValue) {
-                                 if (log.isTraceEnabled())
-                                    log.tracef("Discard key for segment completion %s", keyForSegmentCompletion);
-                                 keyForSegmentCompletion = null;
-                              }
-                           }
-                        });
+                  Flowable<SegmentAwarePublisherSupplier.NotificationWithLost<Object>> mappedFlowable =
+                        Flowable.fromPublisher(functionToApply.apply(Flowable.just(originalValue)))
+                              .map(v -> Notifications.value(v, notification.valueSegment()));
+                  Object key = toKeyFunction.apply(originalValue);
+                  return mappedFlowable
+                        // Signal the end of the key - flatMap could have 0 or multiple entries
+                        .concatWith(Single.just(new KeyCompleted<>(key, notification.valueSegment())));
```

Comment:
> And I wish we could avoid this transformer invocation for each value here, too.

This was not changed btw, it was this way before. I am not aware of a way in a memory/streaming fashion to do this otherwise. It may be possible, but I would have to think about what we can reduce there. The problem is you have to keep track of what key maps to what transformed value.

> Reading more on that, one thing I didn't quite understand is that it seems KeyPublisherState is only used when the stream has filterKeys(keySet). Why do we need it when we are processing only a (small, in theory) set of keys, but not when we are processing all the keys in the cache? Or did I miss something and KeyPublisherState is also used when keysToInclude == null?

This is not the case and actually has nothing to with the context. KeyPublisherState is used whenever you have a key tracking operation (such as iteration or forEach in DistributedCacheStream terms) when there is also a map or flatMap operation applied and EXACTLY_ONCE guarantees are applied. We have to track what transformed value(s) map to a given key. https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManagerImpl.java#L1165